vignettes/special-files.Rmd

```diff
@@ -0,0 +1,77 @@
+---
+title: "Special files"
+output: rmarkdown::html_vignette
+vignette: >
+  %\VignetteIndexEntry{Special files}
+  %\VignetteEngine{knitr::rmarkdown}
+  %\VignetteEncoding{UTF-8}
+---
+
+```{r, include = FALSE}
+knitr::opts_chunk$set(
+  collapse = TRUE,
+  comment = "#>"
+)
+```
+
+```{r setup}
+#| include: false
+library(testthat)
+```
+
+This vignette describes the various special files that testthat works with: test, helper, and setup files.
+
+## Test files
+
+These are bread and butter of testthat.
+Test files live in `tests/testthat`, start with either `test-` or `test_`, and end with `.r` or `.R`.
+We recommend organising your test files so that there's a one-to-one correspondence between the files in `R/` and the files in `tests/testthat/` so that (e.g.) `R/file-one.R` has a matching `tests/testthat/test-one.R`.
+This correspondence is maintained by functions like `usethis::use_r()` and `usethis::use_test()` and by respected by functions like `devtools::test_active_file()` and `devtools::test_coverage_active_file()`.
+
+Test files are executed in alphabetical order, but you should strive to avoid dependencies between test files.
+
+## Helper files
+
+Helper files live in `tests/testtthat`, start with `helper`, and end with `.r` or `.R`.
+They are sourced by `devtools::load_all()` (so they're available interactively when developing your packages) and by `test_check()` and friends (so that they're available no matter how your tests are executed).
+
+Helper files are a useful place for functions that you've extracted from repeated code in your tests, whether that be test fixtures (`vignette("test-fixtures")`), custom expectations (`vignette("custom-expectations")`), or skip helpers (`vignette("skipping")`).
+
+## Setup files
+
+Setup files live in `tests/testthat`, start with `setup`, and end with `.r` or `.R`.
+Typically there is only one setup function which by convention is placed in `tests/testthat/setup.R`.
+Set up files are sourced by `test_check()` and friends (so that they're available no matter how your tests are executed), but they are *not* sourced by `devtools::load_all()`.
+
+Setup files are good place to put truly global test setup that would be impractical to build into every single test and that might be tailored for test execution in non-interactive or remote environments.
+Examples:
+
+-   Turning off behaviour aimed at an interactive user, such as messaging or writing to the clipboard.
+-   Loading a credential for auth.
+-   Setting up a cache folder.
+
+If any of your setup should be reversed after test execution (i.e. it needs to be torn down), we recommend maintaining that teardown code alongside the setup code, in `setup.R`, because this makes it easier to ensure they stay in sync.
+The artificial environment `teardown_env()` exists as a magical handle to use in `withr::defer()` and `withr::local_*()` / `withr::with_*()`.
```

Comment:
Would `with_*()` actually ever be what you want re: `teardown_env()`? I don't really see how that would work.